components/automate-cli/pkg/reportgenerator/usagegenerator.go

```diff
@@ -251,3 +263,257 @@ func queryElasticSearchNodeReport(client *elastic.Client, startTime time.Time, e
+	if err != nil {
+		fmt.Println("Failed to open Compliance scan detail report")
+		os.Exit(1)
+	}
+	defer f.Close()
+
+	w := gocsv.DefaultCSVWriter(f)
+
+	var sourceField = []string{
+		"node_uuid",
+		"version",
+		"node_name",
+		"environment",
+		"platform.name",
+		"controls_sums.total",
+		"controls_sums.passed.total",
+		"controls_sums.skipped.total",
+		"controls_sums.failed.total",
+		"profiles.name",
+		"profiles.title",
+		"profiles.version",
+		"end_time",
+	}
+
+	t := endTime.Add(24 * time.Hour)
+
+	if t.Before(startTime) {
+		t = startTime
+	}
+	header := true
+	for {
+		rangeQuery := elastic.NewRangeQuery("end_time").
+			Format("yyyy-MM-dd||yyyy-MM-dd-HH:mm:ss||yyyy-MM-dd'T'HH:mm:ssZ")
+		rangeQuery.Gte(t)
+		rangeQuery.Lte(endTime)
+
+		fetchSource := elastic.NewFetchSourceContext(true).Include(sourceField...)
+
+		searchService := client.Search().
+			Query(rangeQuery).
+			Index("comp-7-r-*").
+			Size(10000).
+			FetchSourceContext(fetchSource)
+
+		searchResult, err := searchService.Do(context.Background())
+		if err != nil {
+			fmt.Println("Error in query: ", err)
+			os.Exit(1)
+		}
+
+		if searchResult.Hits.TotalHits > 0 {
+			if header == true {
+				headers := []string{"Node_ID", "Version", "Node_Name", "Environment", "End_Time", "Platform__Name", "Controls_Sums__Total", "Controls_Sums__Passed", "Controls_Sums__Skipped", "Controls_Sums__Failed", "Profiles__Count"}
+				if err := w.Write(headers); err != nil {
+					fmt.Println("Error while writing csv: ", err)
+				}
+				w.Flush()
+				if err := w.Error(); err != nil {
+					fmt.Println("Error while writing csv: ", err)
+				}
+			}
+
+			for _, hit := range searchResult.Hits.Hits {
+
+				var payload []byte
+
+				payload = []byte("[")
+				comma := []byte(",")
+				payload = append(payload, *hit.Source...)
+				payload = append(payload, comma...)
+				payload = payload[:len(payload)-1]
+				payload = append(payload, []byte("]")...)
+				var s ComplianceInfo
+
+				err = json.Unmarshal(payload, &s)
+				if err != nil {
+					fmt.Println("Json marshal err :", err)
+					os.Exit(1)
+				}
+
+				for i := 0; i < len(s); i++ {
+					record := []string{
+						s.mapAt(i)["node_uuid"].(string),
+						s.mapAt(i)["version"].(string),
+						s.mapAt(i)["node_name"].(string),
+						s.mapAt(i)["environment"].(string),
+						s.mapAt(i)["end_time"].(string),
+						s.mapAt(i)["platform"].(map[string]interface{})["name"].(string),
+						fmt.Sprint(s.mapAt(i)["controls_sums"].(map[string]interface{})["total"]),
+						fmt.Sprint(s.mapAt(i)["controls_sums"].(map[string]interface{})["passed"].(map[string]interface{})["total"]),
+						fmt.Sprint(s.mapAt(i)["controls_sums"].(map[string]interface{})["skipped"].(map[string]interface{})["total"]),
+						fmt.Sprint(s.mapAt(i)["controls_sums"].(map[string]interface{})["failed"].(map[string]interface{})["total"]),
+						strconv.Itoa(len(s.mapAt(i)["profiles"].([]interface{}))),
+					}
```

Comment:
The whole for loop can be written as a separate function that in turn returns an array of string.